From 6878804af61ce8a583f760b1faa1c3b1bd79efab Mon Sep 17 00:00:00 2001
From: Jordan Ngako <jordanfalken@gmx.de>
Date: Sun, 4 Jan 2026 21:38:40 +0100
Subject: [PATCH] export symbols rockchip iommu

---
 drivers/iommu/rockchip-iommu.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/iommu/rockchip-iommu.c b/drivers/iommu/rockchip-iommu.c
index 4c7f470a4..0dd6cbfb8 100644
--- a/drivers/iommu/rockchip-iommu.c
+++ b/drivers/iommu/rockchip-iommu.c
@@ -25,6 +25,7 @@
 #include <linux/pm_runtime.h>
 #include <linux/slab.h>
 #include <linux/spinlock.h>
+#include <soc/rockchip/rockchip_iommu.h>
 
 #include "iommu-pages.h"
 
@@ -1300,6 +1301,18 @@ static int rk_iommu_probe(struct platform_device *pdev)
 	return err;
 }
 
+bool rockchip_iommu_is_enabled(struct device *dev)
+{
+	struct rk_iommu *iommu;
+
+	iommu = rk_iommu_from_dev(dev);
+	if (!iommu)
+		return false;
+
+	return true;
+}
+EXPORT_SYMBOL(rockchip_iommu_is_enabled);
+
 static void rk_iommu_shutdown(struct platform_device *pdev)
 {
 	struct rk_iommu *iommu = platform_get_drvdata(pdev);
-- 
2.52.0

